[dbsync] Fix open issues in rsync library and make it functional

AchuthaSourabhC opened the pull request:
This PR fixes all the open issues to get the library functional

**Issues being fixed**
1. First sync  failed when file is not present in target, we have to create and directly copy the file first time 
2. Exception was thrown if file size was less than min file size constant, in this case file can just be copied directly
3. Staging files (checksum, instruction,staged) were not stored in proper directory structure and would cause problems when syncing multiple files
4. once rsycn was done and staged file is created based on instructions it was note being moved/copied to target and target file was left out of data
5. Staging files were not getting cleaned up at the end
6. If trailing slash was not present in staging path invalid staging file paths were getting created. ( need to add if missing)
7. Cloud run job were not deleted at the end

**Other changes**

1. Added a param to decide if staging files should be deleted (its helpful to not delete for debugging and testing)
2. Fixed formatting and naming issues ( lot of the updates are related to this)

Testing:
test manually with following params
```
--project=achuthasourabh-test --location=us-west1 --target_file=gs://rsync_target/data/test.csv --source_file=file:///usr/local/google/home/achuthasourabh/Downloads/annual-enterprise-survey-2023-financial-year-provisional-size-bands.csv --staging_bucket=gs://rsync_staging --delete_staging_files=false
```

Screenshot of staging bucket files
![image](https://github.com/user-attachments/assets/a9a51d7a-3eb4-45bb-a392-e37aebdd09a5)

Screenshot of target file
![image](https://github.com/user-attachments/assets/e431f37d-2f5e-4197-adef-aa0f727e6872)

Note: Need to setup proper integ/e2e tests but for now trying to get it working and unblock dependent work

tonyxiaowei commented on dbsync/server/src/main/java/com/google/edwmigration/dbsync/server/RsyncServer.java:
```diff
@@ -10,41 +10,49 @@
+              // TODO: Switch to simple write to save storage and read by byte size
+              checksum.writeDelimitedTo(checksumStream);
+            } catch (IOException e) {
+              throw new RuntimeException("Failed to generate checksum", e);
```
We won't delete the staging files in this case?

AchuthaSourabhC replied:
Similar to other comment https://github.com/google/dwh-migration-tools/pull/782#discussion_r1999107177
Need a good way cleanup regardless of the error

tonyxiaowei commented on dbsync/server/src/main/java/com/google/edwmigration/dbsync/server/GCSTarget.java:
```diff
@@ -55,4 +63,14 @@ public ByteSource getInstructionsByteSource() {
   }
+
+  @Override
+  public void moveStagedFileToTarget(boolean deleteStagingFiles) {
```
I think it's better to have a deleteStagedFiles method instead of passing in boolean flag here.

AchuthaSourabhC replied:
Ya make sense. Will update

AchuthaSourabhC commented on dbsync/client/src/main/java/com/google/edwmigration/dbsync/client/RsyncClient.java:
```diff
@@ -73,18 +87,22 @@ private void putRsync(ByteSource source, CloudRunServerAPI server, RsyncTarget t
+            try {
+              instruction.writeDelimitedTo(instructionStream);
+            } catch (IOException e) {
+              throw new RuntimeException("Failed to write instructions", e);
```
I was thinking whats best way to handle this because failure can happen at different stages and may be some of the files we are deleting are not even created.
One option is we do cleanup on best effort in a finally block at top level that encompasses all exceptions. But ignore any error while trying to delete including file not found errors.

AchuthaSourabhC commented on dbsync/client/src/main/java/com/google/edwmigration/dbsync/client/RsyncClient.java:
```diff
@@ -73,18 +87,22 @@ private void putRsync(ByteSource source, CloudRunServerAPI server, RsyncTarget t
     server.reconstruct();
 
-    //TODO delete the jobs from cloudrun
+    server.deleteJob(Mode.GENERATE);
```
To my knowledge we wanted to run it as jobs running on demand, since frequency of execution might actually be less. 
Though i was checking if there was a way to re-use deployed jobs and run with different inputs instead of deploying new jobs for each file. But for now couldn't find good solution for this.

AchuthaSourabhC commented on dbsync/server/src/main/java/com/google/edwmigration/dbsync/server/CloudRunServerAPI.java:
```diff
@@ -89,12 +98,12 @@ public void deployRsyncJobs()
-        project, location, Mode.GENERATE,
-        String.format("%s && %s", jarDownloadCommand, generateCommand)
-    );
+        project,
```
Yes will keep in mind. I initially had setup "format on save" and didn't realize would end up with big diff

AchuthaSourabhC commented on dbsync/server/src/main/java/com/google/edwmigration/dbsync/server/GCSTarget.java:
```diff
@@ -2,28 +2,36 @@
+    //  If target files path is gs://target_bucket/dir/target/file.txt
+    //  Staging bucket is gs://target_bucket/
+    //  Then staging files will be stored as gs://target_bucket/dir/target/file.txt/staged
+    String targetRelativePath = UriUtil.getRelativePath(targetUri);
```
Yes will update

AchuthaSourabhC commented:
Closing this PR. The other PR raised by @tonyxiaowei  https://github.com/google/dwh-migration-tools/pull/796 will be addressing most of the concerns. This is not needed anymore
